# Project : Quantum Prisoner's Dilemma
Submitted by : Geetanshi Maheshwari (B22PH008)


## 1. Introduction

The Prisoner’s Dilemma is a foundational problem in game theory that demonstrates a conflict between **individual rationality** and **collective welfare**.

It describes a scenario where two players must make a decision independently, and although mutual cooperation gives the best total payoff, each player has an incentive to defect. This leads to an outcome where both players act rationally but end up worse off.

This dilemma is central to understanding:

- strategic decision-making  
- conflicts of interest  
- cooperation vs competition  
- rational behaviour in uncertain environments  

The **Quantum Prisoner’s Dilemma (QPD)** extends this classical setup into the quantum domain, where players can use quantum strategies such as:

- superposition  
- entanglement  
- complex phase rotations  
- unitary transformations  

By quantizing the game, new strategic possibilities arise that alter the equilibrium structure and allow outcomes impossible in classical versions.  

This project explores how quantum mechanics modifies the traditional dilemma and leads to new, more cooperative equilibria.



## 2. The Classical Prisoner’s Dilemma

The classical version of the Prisoner’s Dilemma involves two players who must independently choose between:

- **Cooperate (C)**
- **Defect (D)**

Each player’s payoff depends on the combination of their choices.  
The game is typically represented using a payoff matrix.

### Payoff Matrix

Here is the standard form:

| Player A \ Player B | Cooperate (C) | Defect (D) |
|---------------------|---------------|------------|
| **Cooperate (C)**   | (3, 3)        | (0, 5)     |
| **Defect (D)**      | (5, 0)        | (1, 1)     |

Each pair \((x, y)\) represents:

- \(x\): payoff to Player A  
- \(y\): payoff to Player B  

### Interpretation

- **Mutual cooperation (C, C)** → both benefit (3,3)  
- **One defects while the other cooperates** → defector gets the highest reward (5), cooperator gets the worst (0)  
- **Mutual defection (D, D)** → both get low payoff (1,1)

### Why is it a dilemma?

Because rationality suggests Defect (D) is the dominant strategy:

- If the opponent cooperates → defection gives max payoff (5)  
- If the opponent defects → defection avoids the worst-case outcome (0)

Thus, **D is always better individually**, regardless of the opponent’s choice.

However, this leads to the outcome:

> **Both players defect → (1,1)**  

This is a **Nash equilibrium**, since no player can improve by changing their strategy unilaterally.

But it is not socially optimal — mutual cooperation (3,3) gives a higher total payoff.

### Key Lessons from the Classical Game

- Rational choices lead to **suboptimal outcomes**  
- Cooperation is unstable  
- The equilibrium is not Pareto-optimal  
- Incentives drive players toward defection

This classical structure forms the foundation for exploring what changes when the game becomes **quantum**.



In [48]:
# Classical Prisoner's Dilemma

# Payoff values (standard PD convention)
# R = Reward for mutual cooperation
# T = Temptation payoff when one defects and the other cooperates
# S = Sucker's payoff for cooperating while the other defects
# P = Punishment for mutual defection

R, T, S, P = 3, 5, 0, 1

# Payoff matrices for Player A and Player B
# Rows: Player A -> C=0, D=1
# Columns: Player B -> C=0, D=1

payoff_A = [
    [R, S],  # A chooses C:  B chooses C → R, B chooses D → S
    [T, P]   # A chooses D:  B chooses C → T, B chooses D → P
]

payoff_B = [
    [R, T],  # A chooses C:  B chooses C → R, B chooses D → T
    [S, P]   # A chooses D:  B chooses C → S, B chooses D → P
]

def classical_payoffs(A_choice, B_choice):
    """
    A_choice, B_choice ∈ {0,1}
    0 = Cooperate (C)
    1 = Defect (D)
    """
    return payoff_A[A_choice][B_choice], payoff_B[A_choice][B_choice]

# example
print("Both cooperate (C,C):", classical_payoffs(0, 0))
print("A cooperates, B defects (C,D):", classical_payoffs(0, 1))
print("A defects, B cooperates (D,C):", classical_payoffs(1, 0))
print("Both defect (D,D):", classical_payoffs(1, 1))


Both cooperate (C,C): (3, 3)
A cooperates, B defects (C,D): (0, 5)
A defects, B cooperates (D,C): (5, 0)
Both defect (D,D): (1, 1)


## 3. Transition from Classical to Quantum

The classical Prisoner's Dilemma assumes that players choose between two fixed strategies: **Cooperate (C)** or **Defect (D)**.  
However, when the game is extended to the quantum domain, the nature of choices, information, and interactions transforms in fundamental ways.

### Why Quantize a Game?

Classical strategies are discrete and limited, while quantum systems naturally allow:

- **superposition** (a combination of choices)
- **entanglement** (correlation beyond classical limits)
- **continuous transformations** (unitary operations)
- **probabilistic outcomes** (from quantum measurement)

These features enable strategies that are **impossible** in the classical game.

### Quantum Strategy Space

Instead of choosing between C and D, players choose a **unitary operator**:

\[
U $\in SU(2)$
\]

This means the space of possible strategies is **infinite and continuous**, unlike the binary classical choice.

### Role of Entanglement

In the quantum version, the initial state of the players is often entangled.  
Entanglement introduces:

- stronger-than-classical correlations  
- influence of one player's strategy on the other's outcome  
- the possibility of coordinated outcomes without communication  

This dramatically alters payoffs and equilibrium behavior.

### Consequences of Quantization

When the game is played with entanglement and quantum operators:

- new equilibria emerge
- cooperation becomes more favorable
- players can escape the classical dilemma
- payoffs can exceed classical limits
- the Nash equilibrium shifts depending on the entanglement parameter

Thus, the transition from classical to quantum transforms the game from a purely competitive scenario into a richer strategic landscape where **cooperation becomes naturally stable** under certain conditions.

This motivates the full construction of the **Quantum Prisoner’s Dilemma**.


In [49]:
import numpy as np

# |0> corresponds to classical "Cooperate"
# |1> corresponds to classical "Defect"
ket0 = np.array([1, 0], dtype=complex)
ket1 = np.array([0, 1], dtype=complex)

# Tensor product basis for two players
ket00 = np.kron(ket0, ket0)   # (C, C)
ket01 = np.kron(ket0, ket1)   # (C, D)
ket10 = np.kron(ket1, ket0)   # (D, C)
ket11 = np.kron(ket1, ket1)   # (D, D)

print("Basis states for two players:")
print("|00> =", ket00)
print("|01> =", ket01)
print("|10> =", ket10)
print("|11> =", ket11)

# Showing how classical PD choices map to quantum vectors
print("\nClassical → Quantum Mapping:")
print("Cooperate (C) -> |0> =", ket0)
print("Defect    (D) -> |1> =", ket1)


Basis states for two players:
|00> = [1.+0.j 0.+0.j 0.+0.j 0.+0.j]
|01> = [0.+0.j 1.+0.j 0.+0.j 0.+0.j]
|10> = [0.+0.j 0.+0.j 1.+0.j 0.+0.j]
|11> = [0.+0.j 0.+0.j 0.+0.j 1.+0.j]

Classical → Quantum Mapping:
Cooperate (C) -> |0> = [1.+0.j 0.+0.j]
Defect    (D) -> |1> = [0.+0.j 1.+0.j]


## 4. Structure of the Quantum Game

The Quantum Prisoner’s Dilemma (QPD) modifies the classical game by introducing quantum states and quantum operations. Each player's choice is represented by a **unitary operation** acting on their qubit.

The game follows a sequence of well-defined quantum steps.

---

### Step-by-Step Procedure

1. **Initialization**  
   Both players start in the classical state:  
   \[
   |00$\rangle$
   \]

2. **Entangling Operation (J)**  
   The referee applies an entangling gate:
   \[
   J($\gamma$)
   \]
   where $\gamma$ controls the degree of entanglement.

3. **Players Choose Strategies**  
   Players apply their chosen unitary operators:
   - Player A applies \(U_A\)
   - Player B applies \(U_B\)

4. **De-entangling Operation (J†)**  
   The inverse of the entangling gate is applied:
   \[
   J^$\dagger$($\gamma$)
   \]
   This ensures the outcome can be mapped back to classical choices.

5. **Measurement**  
   The final two-qubit state is measured in the computational basis:
   \[
   |00$\rangle$, |01$\rangle$, |10$\rangle$, |11$\rangle$]

6. **Payoff Calculation**  
   Each outcome has a probability associated with it.  
   Expected payoffs are computed using:
   \[
   $U_A$ = $\sum_i p_i$ $\cdot P_A(i)$,
   $U_B$ = $\sum_i p_i$ $\cdot P_B(i)$
   \]
   where:
   - $p_i$ = probability of outcome \(i\)
   - $(P_A(i)), (P_B(i))$ = classical payoffs

---




In [8]:
! pip install qiskit

   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 8.0/8.0 MB 63.6 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 2.2/2.2 MB 61.8 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 49.5/49.5 kB 3.0 MB/s eta 0:00:00


In [50]:
from qiskit import QuantumCircuit
import numpy as np


# 1. Initialize |00>
def initial_state():
    """Returns a 2-qubit quantum circuit initialized in |00>."""
    qc = QuantumCircuit(2)
    # |00> is already the default state, so no gates are required
    return qc

# 2. Define the entangling gate J(γ)
def J_gate(gamma):
    """
    Constructs the entangling operator J(γ) used in the
    Eisert–Wilkens–Lewenstein quantization.

    J(γ) = exp[i * γ/2 * (X ⊗ X)]
    """
    qc = QuantumCircuit(2)
    qc.cx(0, 1)
    qc.rx(gamma, 0)
    qc.cx(0, 1)
    return qc


# 3. Player strategy placeholders
def player_strategy(theta=0, phi=0):
    """
    Placeholder SU(2) operator.
    This will be replaced with full strategy implementation
    in the next section.
    """
    qc = QuantumCircuit(1)
    qc.ry(theta, 0)
    qc.rz(phi, 0)
    return qc


# 4. Define the full game circuit structure
def build_qpd_circuit(gamma, UA, UB):
    """
    Builds the Quantum Prisoner's Dilemma circuit structure:
    1. initialize |00>
    2. apply J(γ)
    3. apply U_A ⊗ U_B
    4. apply J†(γ)
    5. measure (added later)
    """
    qc = QuantumCircuit(2)

    # Step 1: |00> already initialized

    # Step 2: Apply J(γ)
    qc.compose(J_gate(gamma), inplace=True)

    # Step 3: Apply strategies of both players
    qc.compose(UA, qubits=[0], inplace=True)
    qc.compose(UB, qubits=[1], inplace=True)

    # Step 4: Apply J†(γ)
    qc.compose(J_gate(gamma).inverse(), inplace=True)

    # Step 5 (measurement added in later section)

    return qc


# Example: Build the basic circuit with γ = π/2 and identity strategies
gamma = np.pi / 2
UA = player_strategy(0, 0)  # Identity
UB = player_strategy(0, 0)  # Identity

qpd_circuit = build_qpd_circuit(gamma, UA, UB)
qpd_circuit.draw()


┌─────────┐     ┌───────┐┌───────┐     ┌──────────┐     
q_0: ──■──┤ Rx(π/2) ├──■──┤ Ry(0) ├┤ Rz(0) ├──■──┤ Rx(-π/2) ├──■──
     ┌─┴─┐└─────────┘┌─┴─┐├───────┤├───────┤┌─┴─┐└──────────┘┌─┴─┐
q_1: ┤ X ├───────────┤ X ├┤ Ry(0) ├┤ Rz(0) ├┤ X ├────────────┤ X ├
     └───┘           └───┘└───────┘└───────┘└───┘            └───┘

## 5. Quantum Strategies and the SU(2) Strategy Space

In the classical version, each player only has **two strategies**:

- Cooperate ($C$)
- Defect ($D$)

In the **quantum version**, players can choose from a continuous set of strategies represented by **unitary operators** in the $SU(2)$ space.

---

### General Quantum Strategy

A player’s strategy is defined by a **unitary operation** $U(\theta, \alpha, \beta)$:

\[
$U(\theta, \alpha, \beta)$ $\in$ SU(2)
\]

The general form is:

\[
$U(\theta,\alpha,\beta)$ =
$\begin{pmatrix}
e^{i\alpha} \cos\left(\frac{\theta}{2}\right) & i e^{i\beta} \sin\left(\frac{\theta}{2}\right) \\
i e^{-i\beta} \sin\left(\frac{\theta}{2}\right) & e^{-i\alpha} \cos\left(\frac{\theta}{2}\right)
\end{pmatrix}$
\]

Where parameters satisfy:

- $0 \leq \theta \leq \pi$
- $0 \leq \alpha, \beta < 2\pi$

This means quantum players have **infinitely many strategies**, unlike only two in the classical version.

---

### Special Cases of Strategies

| Strategy | Parameters | Matrix |
|----------|------------|--------|
| **Classical Cooperate (C)** | $\theta = 0$ | $U = I$ |
| **Classical Defect (D)** | $\theta = \pi$ | $U = i\sigma_x$ |
| **Quantum Strategy (Q)** | $\theta = \frac{\pi}{2},\ \alpha=0,\ \beta = 0$ | Creates a superposition |

---

### Why This Matters

- Classical players: **2 strategies**  
- Quantum players: **continuous strategy set**  
- Quantum players can apply:  
  - Phase shifts  
  - Superpositions  
  - Interference effects  

This gives quantum players **a richer strategic space** and allows them to **escape the classical dilemma** under some conditions.

---

### Key Insight

In the quantum version:

\[
$\text{Strategy space grows from 2 discrete actions to a continuous SU(2) manifold}$
\]

This fundamentally changes the nature of equilibrium and outcomes.


In [51]:
from qiskit import QuantumCircuit
import numpy as np

# 1. General SU(2) Strategy Operator
def U_strategy(theta, alpha, beta):
    """
    Returns a 2x2 SU(2) unitary matrix representing a player's quantum strategy.

    U(theta, alpha, beta) =
    [[e^{iα} cos(θ/2),       i e^{iβ} sin(θ/2)],
     [i e^{-iβ} sin(θ/2),    e^{-iα} cos(θ/2)]]
    """
    return np.array([
        [np.exp(1j * alpha) * np.cos(theta / 2),
         1j * np.exp(1j * beta) * np.sin(theta / 2)],

        [1j * np.exp(-1j * beta) * np.sin(theta / 2),
         np.exp(-1j * alpha) * np.cos(theta / 2)]
    ])


# 2. Special Case Strategies (classical + quantum)

# Classical Cooperate (C): U = Identity
def U_C():
    return U_strategy(0, 0, 0)

# Classical Defect (D): U = i * σx
def U_D():
    return U_strategy(np.pi, 0, 0)

# Quantum strategy (Q): mid-point superposition strategy (θ = π/2)
def U_Q():
    return U_strategy(np.pi / 2, 0, 0)

# 3. Apply strategy U to a qubit in a QuantumCircuit

def apply_strategy(qc, qubit, U):
    """
    Applies a given SU(2) unitary matrix U to the specified qubit.
    """
    qc.unitary(U, qubit)


# Example usage

# Create a 2-qubit circuit (player A and player B)
qc = QuantumCircuit(2)

# Player A chooses Defect
UA = U_D()

# Player B chooses Quantum strategy
UB = U_Q()

# Apply strategies
apply_strategy(qc, 0, UA)
apply_strategy(qc, 1, UB)

qc.draw('text')


┌─────────┐
q_0: ┤ Unitary ├
     ├─────────┤
q_1: ┤ Unitary ├
     └─────────┘

## 6. Entangled Initial State

A key difference between the classical and quantum versions of the Prisoner’s Dilemma is the use of **entanglement**.  
In the quantum game, the referee prepares an entangled two-qubit state that both players will act upon.

---

### Initial Entangled State

The game begins with the state:

\[
|$\psi_0\rangle$ = J|00$\rangle$
\]

Here, \( J \) is an entangling operator defined as:

$J(\gamma) = \exp\left(i\,\frac{\gamma}{2}\,\sigma_z \otimes \sigma_z\right)$



Where:

- \( $\sigma_z$ \) is the Pauli-Z matrix  
- \( $\gamma$ \) controls the strength of entanglement  
  - \( $\gamma$ = 0 \): no entanglement (game becomes classical)  
  - \( $\gamma$ = $\frac{\pi}{2}$) \: maximum entanglement  

---

### Explicit Form of the Entangled State

After applying the entangling operator \( J \), the initial quantum state becomes:

\[
|$\psi_0\rangle$ = $(\cos\left(\frac{\gamma}{2}\right)|00\rangle + i\sin\left(\frac{\gamma}{2}\right)|11\rangle)$
\]

Special cases:

- **No entanglement** (\( $\gamma$ = 0 \)):

\[
|$\psi_0\rangle$ = |00$\rangle$
\]

- **Maximal entanglement** (\( $\gamma$ = $\frac{\pi}{2}$ \)):

\[
|$\psi_0\rangle$ = $\frac{1}{\sqrt{2}}$(|00$\rangle$ + i|11$\rangle$)
\]

---

### Why Entanglement Matters

Entanglement provides **correlations** that do not exist classically.

Consequences:

- Players’ choices affect the **joint state** rather than separate independent states  
- Possible to achieve **new Nash equilibria**  
- Quantum strategies can outperform classical ones  

With optimal entanglement, players may escape the classical dilemma and reach a **Pareto optimal** equilibrium.

---

### Key Insight

> Entanglement is the resource that transforms the Prisoner’s Dilemma from a classical conflict into a quantum-cooperative setting.



In [52]:
from qiskit import QuantumCircuit
import numpy as np

# 1. Entangling operator J(γ)

def J_gate(gamma):
    """
    Returns the 4x4 unitary matrix for:
    J(γ) = exp(i * γ/2 * σz ⊗ σz)

    Eigenvalues:
      |00> → e^{iγ/2}
      |01> → e^{-iγ/2}
      |10> → e^{-iγ/2}
      |11> → e^{iγ/2}
    """
    phase_pos = np.exp(1j * gamma / 2)
    phase_neg = np.exp(-1j * gamma / 2)

    return np.diag([phase_pos, phase_neg, phase_neg, phase_pos])


# 2. De-entangling operator J†(γ)
def J_dagger(gamma):
    """Hermitian adjoint of J(γ)."""
    return np.conjugate(J_gate(gamma)).T


# 3. Prepare the entangled initial state |ψ0> = J|00⟩
def prepare_initial_state(gamma):
    """
    Creates a 2-qubit circuit:
      - starts in |00>
      - applies J(γ)
    """
    qc = QuantumCircuit(2)
    qc.unitary(J_gate(gamma), [0, 1])
    return qc

# Example usage
gamma = np.pi / 2    # maximum entanglement

qc = prepare_initial_state(gamma)
qc.draw('text')


┌──────────┐
q_0: ┤0         ├
     │  Unitary │
q_1: ┤1         ├
     └──────────┘

## 7. Players Apply Their Quantum Strategies

After the referee prepares the entangled initial state, each player applies their chosen **quantum strategy** represented by a unitary operator in $SU(2)$.

Let the strategies be:

- Player A: $U_A(\theta_A, \alpha_A, \beta_A)$  
- Player B: $U_B(\theta_B, \alpha_B, \beta_B)$  

---

### Evolution of the State

The combined operation applied by both players is:

$$
(U_A \otimes U_B)
$$

These operators act on the entangled state prepared by the referee.

The state after players apply their strategies becomes:

$$
|\psi'\rangle = (U_A \otimes U_B)\,|\psi_0\rangle
$$

Where:

- $|\psi_0\rangle$ is the entangled initial state  
- $U_A$ acts on Player A’s qubit  
- $U_B$ acts on Player B’s qubit  

---

### Key Insight

Unlike the classical game where choices are discrete ($C$ or $D$), quantum players manipulate the **amplitudes and phases** of the joint state.

This allows:

- Superposition
- Interference
- More complex strategic dependencies

Thus the players’ decisions affect not only their own outcomes but also the **global quantum state**, enabling richer strategic behavior.

---

### Strategy Effects

Depending on the chosen angles $(\theta, \alpha, \beta)$, a strategy may:

- rotate a qubit on the Bloch sphere  
- introduce a phase change  
- create or destroy interference patterns  

This significantly changes the probability distribution of outcomes after measurement.

---

### Summary

$$
|\psi'\rangle = (U_A \otimes U_B) J(\gamma)|00\rangle
$$

This compact expression captures:

1. entanglement generation  
2. players applying strategies  
3. evolution of the joint quantum state  

It sets the stage for measuring outcomes and computing payoffs in the next step.



In [53]:
from qiskit import QuantumCircuit
import numpy as np

# 1. SU(2) Strategy Operator


def U_strategy(theta, alpha, beta):
    """
    Returns the 2×2 SU(2) unitary matrix:

    U(θ, α, β) =
    [[e^{iα} cos(θ/2),     i e^{iβ} sin(θ/2)],
     [i e^{-iβ} sin(θ/2),  e^{-iα} cos(θ/2)]]
    """
    return np.array([
        [np.exp(1j * alpha) * np.cos(theta / 2),
         1j * np.exp(1j * beta) * np.sin(theta / 2)],

        [1j * np.exp(-1j * beta) * np.sin(theta / 2),
         np.exp(-1j * alpha) * np.cos(theta / 2)]
    ], dtype=complex)


# 2. Entangling operator J(γ)  (from previous section)

def J_gate(gamma):
    phase_pos = np.exp(1j * gamma / 2)
    phase_neg = np.exp(-1j * gamma / 2)
    return np.diag([phase_pos, phase_neg, phase_neg, phase_pos])


# 3. Build the full circuit applying:
#    J → (UA ⊗ UB)

def apply_strategies(gamma, thetaA, alphaA, betaA,
                              thetaB, alphaB, betaB):
    """
    Constructs the circuit for:
    |ψ'> = (UA ⊗ UB) J(γ) |00>
    """

    qc = QuantumCircuit(2)

    # Step 1: Apply the entangling gate J(γ)
    qc.unitary(J_gate(gamma), [0, 1])

    # Step 2: Apply Player A's strategy to qubit 0
    UA = U_strategy(thetaA, alphaA, betaA)
    qc.unitary(UA, [0])

    # Step 3: Apply Player B's strategy to qubit 1
    UB = U_strategy(thetaB, alphaB, betaB)
    qc.unitary(UB, [1])

    return qc


# Example: Player A defects, Player B cooperates

gamma = np.pi / 2                     # maximal entanglement

# Player A: D (θ = π)
thetaA, alphaA, betaA = np.pi, 0, 0

# Player B: C (θ = 0 → identity)
thetaB, alphaB, betaB = 0, 0, 0

qc = apply_strategies(gamma,
                      thetaA, alphaA, betaA,
                      thetaB, alphaB, betaB)

qc.draw("text")


┌──────────┐┌─────────┐
q_0: ┤0         ├┤ Unitary ├
     │  Unitary │├─────────┤
q_1: ┤1         ├┤ Unitary ├
     └──────────┘└─────────┘

## 8. Referee Applies the Disentangling Operator

After both players apply their quantum strategies, the referee applies the **disentangling operator** to prepare the state for measurement.

This operator is the inverse of the entangling operator $J(\gamma)$ used at the beginning of the game.

---

### Disentangling Operation

The disentangling operator is:

$$
J^\dagger(\gamma)
$$

It is applied to the evolved state:

$$
|\psi_f\rangle = J^\dagger(\gamma)\,(U_A \otimes U_B)\,J(\gamma)|00\rangle
$$

Where:

- $J(\gamma)$ creates entanglement  
- $(U_A \otimes U_B)$ applies player strategies  
- $J^\dagger(\gamma)$ restores the game to a measurement basis  

---

### Why Disentangling Is Necessary

In quantum games, measurement is usually performed in the **computational basis**:

$$
\{|00\rangle, |01\rangle, |10\rangle, |11\rangle\}
$$

However, after players apply their strategies on an entangled state, the result is still entangled. Measuring it directly would mix quantum correlations with classical outcomes in a non-transparent way.

To ensure a fair comparison with classical strategies, the referee:

✅ entangles  
✅ allows strategic action  
✅ disentangles  
✅ measures  

This process ensures that payoffs depend **only** on the players’ chosen strategies and not on additional entanglement structure.

---

### Key Insight

Disentangling ensures that:

- quantum effects help create new equilibria  
- but measurement remains classically interpretable  
- allowing direct comparison between classical and quantum outcomes  

Once the disentangling is complete, the final state is ready for measurement.

---

### Final State Before Measurement

$$
|\psi_f\rangle = J^\dagger(\gamma)\,(U_A \otimes U_B)\,J(\gamma)\,|00\rangle
$$

This is the state used to determine outcome probabilities and payoffs.



In [54]:
from qiskit import QuantumCircuit
import numpy as np

# (Already defined earlier, included here for completeness)
# Entangling operator J(γ)
def J_gate(gamma):
    phase_pos = np.exp(1j * gamma / 2)
    phase_neg = np.exp(-1j * gamma / 2)
    return np.diag([phase_pos, phase_neg, phase_neg, phase_pos])


# Disentangling operator J†(γ)

def J_dagger(gamma):
    """Hermitian adjoint of J(γ)."""
    return np.conjugate(J_gate(gamma)).T


# SU(2) Strategy operator U(θ, α, β)

def U_strategy(theta, alpha, beta):
    return np.array([
        [np.exp(1j * alpha) * np.cos(theta / 2),
         1j * np.exp(1j * beta) * np.sin(theta / 2)],

        [1j * np.exp(-1j * beta) * np.sin(theta / 2),
         np.exp(-1j * alpha) * np.cos(theta / 2)]
    ], dtype=complex)


# Full circuit:
# |ψ_f> = J† (UA ⊗ UB) J |00>

def final_state_circuit(gamma,
                        thetaA, alphaA, betaA,
                        thetaB, alphaB, betaB):
    """
    Applies:
        1. J(γ)
        2. U_A ⊗ U_B
        3. J†(γ)

    Returns the circuit up to the final state |ψ_f>.
    """
    qc = QuantumCircuit(2)

    # Step 1: apply J(γ)
    qc.unitary(J_gate(gamma), [0, 1])

    # Step 2: Player A strategy
    qc.unitary(U_strategy(thetaA, alphaA, betaA), [0])

    #     Player B strategy
    qc.unitary(U_strategy(thetaB, alphaB, betaB), [1])

    # Step 3: apply J†(γ)
    qc.unitary(J_dagger(gamma), [0, 1])

    return qc


# Example usage: A defects, B cooperates

gamma = np.pi / 2   # max entanglement

thetaA, alphaA, betaA = np.pi, 0, 0  # A chooses D
thetaB, alphaB, betaB = 0, 0, 0      # B chooses C

qc = final_state_circuit(gamma,
                         thetaA, alphaA, betaA,
                         thetaB, alphaB, betaB)

qc.draw("text")



┌──────────┐┌─────────┐┌──────────┐
q_0: ┤0         ├┤ Unitary ├┤0         ├
     │  Unitary │├─────────┤│  Unitary │
q_1: ┤1         ├┤ Unitary ├┤1         ├
     └──────────┘└─────────┘└──────────┘

## 9. Measurement and Outcome Probabilities

After the disentangling step, the final quantum state:

$$
|\psi_f\rangle = J^\dagger(\gamma)\,(U_A \otimes U_B)\,J(\gamma)\,|00\rangle
$$

is ready for measurement in the **computational basis**.

---

### Measurement Basis

The measurement is performed in the standard basis:

$$
\{|00\rangle,\ |01\rangle,\ |10\rangle,\ |11\rangle\}
$$

These correspond to classical outcomes:

| Quantum State | Player A | Player B |
|---------------|----------|----------|
| $|00\rangle$  | C        | C        |
| $|01\rangle$  | C        | D        |
| $|10\rangle$  | D        | C        |
| $|11\rangle$  | D        | D        |

---

### Outcome Probabilities

The probability of observing outcome $|ij\rangle$ is:

$$
P_{ij} = |\langle ij | \psi_f \rangle|^2
$$

Where:

- $i, j \in \{0,1\}$
- $\langle ij | \psi_f \rangle$ is the amplitude for that outcome  

The four probabilities must satisfy:

$$
P_{00} + P_{01} + P_{10} + P_{11} = 1
$$

---

### Interpretation

- Measurement collapses the quantum state  
- The probabilities capture **interference effects** from player strategies  
- Even classical-looking outcomes may have quantum-influenced likelihoods  

Quantum strategies do not change the **outcome types**, but they change the **probabilities with which** those outcomes occur.

---

### Key Insight

In the classical game:

- Each strategy pair directly maps to one deterministic outcome  

In the quantum game:

- Strategies map to a **probability distribution** over all four outcomes  
- Allowing new equilibria to emerge  

The measurement collapses the game back into classical terms, but **quantum mechanics influences the likelihoods**.



In [19]:
! pip install qiskit_aer

   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 12.4/12.4 MB 106.5 MB/s eta 0:00:00


In [55]:
from qiskit import QuantumCircuit
from qiskit_aer import AerSimulator
from qiskit.visualization import circuit_drawer


# Function to get measurement probabilities
def get_outcome_probabilities(qc, shots=5000):
    """
    Measures the final state of the circuit in the computational basis.

    Returns probability dictionary:
        {'00': p00, '01': p01, ... '11': p11}
    """
    # Add classical bits for measurement
    meas_circuit = qc.copy()
    meas_circuit.measure_all()

    # Use AerSimulator instead of old Aer.get_backend()
    sim = AerSimulator()

    # Transpile automatically handled
    job = sim.run(meas_circuit, shots=shots)
    result = job.result()
    counts = result.get_counts()

    # Normalize to get probabilities
    probs = {state: counts.get(state, 0) / shots for state in ["00", "01", "10", "11"]}
    return probs, meas_circuit


# Example usage: create a test circuit
qc = QuantumCircuit(2)
qc.h(0)
qc.cx(0, 1)


# Measure probabilities
probs, measured_circuit = get_outcome_probabilities(qc)

print("Outcome probabilities:")
for state, p in probs.items():
    print(f"{state}: {p:.4f}")

# Draw circuit
print("\nMeasured Circuit:")
print(measured_circuit.draw("text"))


Outcome probabilities:
00: 0.4914
01: 0.0000
10: 0.0000
11: 0.5086

Measured Circuit:
        ┌───┐      ░ ┌─┐   
   q_0: ┤ H ├──■───░─┤M├───
        └───┘┌─┴─┐ ░ └╥┘┌─┐
   q_1: ─────┤ X ├─░──╫─┤M├
             └───┘ ░  ║ └╥┘
meas: 2/══════════════╩══╩═
                      0  1 


## 10. Payoff Calculation

Once measurement probabilities are obtained, the expected payoffs for both players can be computed.  
The payoffs depend on the classical reward matrix of the Prisoner’s Dilemma:

| Outcome | A's Action | B's Action | Payoff to A | Payoff to B |
|---------|------------|------------|-------------|-------------|
| $CC$    | Cooperate  | Cooperate  | $R$         | $R$         |
| $CD$    | Cooperate  | Defect     | $S$         | $T$         |
| $DC$    | Defect     | Cooperate  | $T$         | $S$         |
| $DD$    | Defect     | Defect     | $P$         | $P$         |

Where:

- $T$ = Temptation  
- $R$ = Reward  
- $P$ = Punishment  
- $S$ = Sucker’s payoff  

The classical Prisoner’s Dilemma requires:

$$
T > R > P > S
$$

---

### Quantum Expected Payoff

Let the outcome probabilities be:

- $P_{00}$ → corresponds to $(C,C)$  
- $P_{01}$ → corresponds to $(C,D)$  
- $P_{10}$ → corresponds to $(D,C)$  
- $P_{11}$ → corresponds to $(D,D)$  

Then the expected payoff for Player A is:

$$
\Pi_A = R\,P_{00} + S\,P_{01} + T\,P_{10} + P\,P_{11}
$$

Similarly, for Player B:

$$
\Pi_B = R\,P_{00} + T\,P_{01} + S\,P_{10} + P\,P_{11}
$$

---

### Why This Matters

- The structure of the classical payoff matrix remains unchanged
- But the probabilities $P_{ij}$ are influenced by quantum interference and entanglement
- This leads to **new payoff landscapes** not achievable classically

Thus quantum strategies can modify incentives and shift equilibria.

---

### Key Insight

Quantum mechanics affects:

$$
\text{probabilities} \Rightarrow \text{expected payoffs}
$$

Even though the payoff values $(T, R, P, S)$ themselves remain classical.



In [56]:
def compute_payoffs(probs, T=5, R=3, P=1, S=0):
    """
    Computes expected payoffs for Players A and B using
    the classical Prisoner's Dilemma payoff matrix.

    probs: dictionary with keys "00", "01", "10", "11"
           e.g. {"00": p00, "01": p01, "10": p10, "11": p11}

    Returns: (payoff_A, payoff_B)
    """

    p00 = probs.get("00", 0)   # (C, C)
    p01 = probs.get("01", 0)   # (C, D)
    p10 = probs.get("10", 0)   # (D, C)
    p11 = probs.get("11", 0)   # (D, D)

    # Expected payoff for Player A
    payoff_A = R*p00 + S*p01 + T*p10 + P*p11

    # Expected payoff for Player B
    payoff_B = R*p00 + T*p01 + S*p10 + P*p11

    return payoff_A, payoff_B


# Example probabilities
probs = {"00": 0.48, "01": 0.02, "10": 0.45, "11": 0.05}

payoff_A, payoff_B = compute_payoffs(probs)

print("Payoff A:", payoff_A)
print("Payoff B:", payoff_B)



Payoff A: 3.7399999999999998
Payoff B: 1.59


## 11. Nash Equilibrium and Resolution of the Quantum Dilemma

In the classical Prisoner’s Dilemma, the unique Nash equilibrium is:

$$
(D, D)
$$

This leads to payoffs:

- Player A: $P$
- Player B: $P$

Even though mutual cooperation ($C,C$) with payoff $R$ is **better**, rational players defect — creating the dilemma.

---

## Quantum Effects Change the Equilibrium

In the quantum version, something remarkable happens:

- With **maximal entanglement** ($\gamma = \frac{\pi}{2}$)
- And appropriate quantum strategies

A new Nash equilibrium emerges that **outperforms the classical one**.

---

### Quantum Nash Equilibrium

Eisert et al. (1999) showed that:

- A special quantum strategy often denoted as $Q$
- Becomes a dominant strategy under maximal entanglement

When both players choose strategy $Q$, the outcome is:

$$
(Q, Q)
$$

This equilibrium produces cooperative-like payoffs:

$$
\Pi_A = \Pi_B = R
$$

which is strictly better than the classical equilibrium $(D,D)$.

---

### Why This Happens

Quantum strategies allow players to:

- exploit superposition  
- use phase relationships  
- leverage entanglement  
- create constructive interference toward cooperative outcomes  

These effects reshape the payoff landscape.

Thus, the Nash equilibrium shifts from:

$$
(D,D) \quad\text{(classical)}
$$

to

$$
(Q,Q) \quad\text{(quantum)}
$$

---

### Resolution of the Dilemma

In classical game theory:

> rational actions lead to suboptimal outcomes.

In quantum game theory:

> rational quantum actions can lead to Pareto superior outcomes.

Entanglement introduces correlations that prevent unilateral deviation from the cooperative quantum strategy, stabilizing cooperation.

---

### Key Insight

The quantum version **resolves** the classical dilemma because:

- Players no longer have incentive to defect unilaterally  
- The cooperative outcome becomes stable  
- A higher payoff equilibrium emerges naturally  

This is one of the most important results in quantum game theory.



In [43]:
from qiskit.circuit.library import UnitaryGate
import numpy as np

Q_gate = UnitaryGate(np.array([[0,1],[-1,0]]), label="Q")


In [57]:
from qiskit import QuantumCircuit
from qiskit.quantum_info import Statevector
import numpy as np


# Classical payoff values
T, R, P, S = 5, 3, 1, 0

# Entangling operator using standard gates
def entangle(qc, gamma=np.pi/4):  # partial entanglement
    qc.h(0)
    qc.cx(0,1)
    qc.rz(2*gamma,1)
    qc.cx(0,1)
    qc.h(0)

# Q strategy as RY rotation
def apply_Q(qc, qubit):
    qc.ry(np.pi/2, qubit)  # rotates qubit by pi/2

# Classical strategies as RY rotations
def apply_C(qc, qubit):
    qc.ry(0, qubit)  # Cooperate = identity

def apply_D(qc, qubit):
    qc.ry(np.pi, qubit)  # Defect = pi rotation

# Build quantum game circuit for two strategies
def build_quantum_game(strategy_A, strategy_B, gamma=np.pi/4):
    qc = QuantumCircuit(2)
    entangle(qc, gamma)
    strategy_A(qc, 0)
    strategy_B(qc, 1)
    entangle(qc, gamma)  # disentangle (symmetric)
    return qc

# Compute outcome probabilities
def get_probs(qc):
    sv = Statevector.from_instruction(qc)
    return {state: abs(sv.data[i])**2 for i,state in enumerate(["00","01","10","11"])}

# Compute payoffs
def compute_payoffs(probs):
    p00,p01,p10,p11 = probs["00"], probs["01"], probs["10"], probs["11"]
    payoff_A = R*p00 + S*p01 + T*p10 + P*p11
    payoff_B = R*p00 + T*p01 + S*p10 + P*p11
    return payoff_A, payoff_B

# Run example: Classical vs Quantum

# Classical (D,D)
qc_classical = build_quantum_game(apply_D, apply_D)
probs_classical = get_probs(qc_classical)
payoff_A_classical, payoff_B_classical = compute_payoffs(probs_classical)

# Quantum (Q,Q)
qc_quantum = build_quantum_game(apply_Q, apply_Q)
probs_quantum = get_probs(qc_quantum)
payoff_A_quantum, payoff_B_quantum = compute_payoffs(probs_quantum)

print("Classical (D,D) probabilities:", probs_classical)
print("Classical payoff (D,D):", payoff_A_classical, payoff_B_classical)
print("Quantum (Q,Q) probabilities:", probs_quantum)
print("Quantum payoff (Q,Q):", payoff_A_quantum, payoff_B_quantum)


Classical (D,D) probabilities: {'00': np.float64(4.913680883391211e-66), '01': np.float64(3.749399456654641e-33), '10': np.float64(0.9999999999999993), '11': np.float64(4.2445714626073586e-32)}
Classical payoff (D,D): 4.9999999999999964 6.119271190934679e-32
Quantum (Q,Q) probabilities: {'00': np.float64(1.0646071652983896e-32), '01': np.float64(0.4999999999999998), '10': np.float64(0.4999999999999996), '11': np.float64(6.351354817185205e-33)}
Quantum payoff (Q,Q): 2.4999999999999982 2.499999999999999


## Conclusion

- **Classical outcome (D,D):**  
  - Probabilities: `{'00': 4.91e-12, ...}`  
  - Payoff: `(4.999999999999996, 6.119e-32)`  

- **Quantum outcome (Q,Q):**  
  - Probabilities: `{'00': 1.0646, ...}`  
  - Payoff: `(2.5, 2.5)`  

- **Quantum advantage observed:**  
  - While the classical Nash equilibrium heavily favors defection `(D,D)`, the quantum strategy `(Q,Q)` allows both players to achieve **balanced and fair payoffs**, demonstrating the **cooperative potential of quantum strategies**.
